### PR TITLE
Adds relayout option to CustomMultiChildLayout.

### PR DESCRIPTION
**Pull request by slightfoot**
## Description

Adds relayout option to CustomMultiChildLayout to match the behaviour of CustomSingleChildLayout and other such delegates.

In short the framework has a behaviour where delegates can opt to provide a listenable to repeat the main function. Clippers can `reclip`. Painters can `repaint` and Custom Layouts `relayout`. This works for `CustomSingleChildLayout`.. except not for `CustomMultiChildLayout`.

This PR resolves this issue by duplicating the listen and relayout functionality from `RenderCustomSingleChildLayoutBox` to `RenderCustomMultiChildLayoutBox`.

## Tests

* Duplicated single-child relayout test to multi-child and verified it functions correctly.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

I don't consider this a breaking change because the additional parameter to `MultiChildLayoutDelegate` is an optional named parameter for the abstract class meaning it wont break out sub-classed constructors.

## Example of usage

```dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

void main() {
  runApp(
    MaterialApp(
      theme: ThemeData(
        primaryColor: Colors.indigo,
        accentColor: Colors.pinkAccent,
      ),
      home: ExampleScreen(),
    ),
  );
}

class ExampleScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Split Bar'),
      ),
      body: SplitBar(
        onWeightChanged: (weight) => print('weight: $weight'),
        left: Container(
          color: Colors.red,
          child: Placeholder(),
        ),
        right: PageView(
          children: <Widget>[
            Container(
              color: Colors.blue,
              child: Placeholder(),
            ),
            Container(
              color: Colors.green,
              child: Placeholder(),
            ),
            Container(
              color: Colors.yellow,
              child: Placeholder(),
            ),
          ],
        ),
      ),
    );
  }
}

class SplitBar extends StatefulWidget {
  const SplitBar({
    Key key,
    this.left,
    this.right,
    this.touchOverlap = 16.0,
    this.initialWeight = 0.333,
    this.onWeightChanged,
  }) : super(key: key);

  final Widget left;
  final Widget right;
  final double touchOverlap;
  final double initialWeight;
  final ValueChanged<double> onWeightChanged;

  @override
  _SplitBarState createState() => _SplitBarState();
}

class _SplitBarState extends State<SplitBar> {
  ValueNotifier<double> _weight;

  @override
  void initState() {
    super.initState();
    _weight = ValueNotifier<double>(widget.initialWeight);
  }

  void _handleDragUpdate(DragUpdateDetails d) {
    final RenderBox container = context.findRenderObject();
    final pos = container.globalToLocal(d.globalPosition);
    _weight.value = pos.dx / container.size.width;
    widget.onWeightChanged?.call(_weight.value);
  }

  @override
  Widget build(BuildContext context) {
    return CustomMultiChildLayout(
      delegate: _SplitBarDelegate(_weight, widget.touchOverlap),
      children: <Widget>[
        LayoutId(id: 0, child: widget.left),
        LayoutId(id: 1, child: widget.right),
        LayoutId(
          id: 2,
          child: GestureDetector(
            behavior: HitTestBehavior.translucent,
            onHorizontalDragUpdate: _handleDragUpdate,
          ),
        ),
      ],
    );
  }
}

class _SplitBarDelegate extends MultiChildLayoutDelegate {
  _SplitBarDelegate(this.weight, this.overlap) : super(relayout: weight);

  final ValueNotifier<double> weight;
  final double overlap;

  @override
  void performLayout(Size size) {
    final double left = size.width * weight.value;
    final double right = size.width - left;
    layoutChild(0, BoxConstraints.tightFor(width: left, height: size.height));
    positionChild(0, Offset.zero);
    layoutChild(1, BoxConstraints.tightFor(width: right, height: size.height));
    positionChild(1, Offset(left, 0.0));
    layoutChild(2, BoxConstraints.tightFor(width: overlap * 2, height: size.height));
    positionChild(2, Offset(left - overlap, 0.0));
  }

  @override
  bool shouldRelayout(_SplitBarDelegate oldDelegate) {
    return oldDelegate.weight.value != weight.value || oldDelegate.overlap != overlap;
  }
}
```

### CODE REVIEW

File: packages/flutter/test/widgets/custom_multi_child_layout_test.dart [hunk 1]
```diff
@@ -161,4 +178,23 @@ void main() {
     expect(box.size.width, equals(350.0));
     expect(box.size.height, equals(250.0));
   });
+
+  testWidgets('Can use listener for relayout', (WidgetTester tester) async {
+    final ValueNotifier<Size> size = ValueNotifier<Size>(const Size(100.0, 200.0));
+
+    await tester.pumpWidget(Center(
```
Comment (chunhtai) [inline packages/flutter/test/widgets/custom_multi_child_layout_test.dart]: nit: center should be on a new line
```dart
tester.pumpWidget(
  Center(
     child: CustomMultiChildLayout(
        delegate: NotifierLayoutDelegate(size),
      ),
   )
);
```

**Comment by slightfoot** [comment]
@goderbauer @chunhtai review changes made.

**Review by chunhtai — APPROVED** [review]
LGTM

**Review by goderbauer — APPROVED** [review]
LGTM with a handful of nits.

**Comment by goderbauer** [comment]
Also, it looks like you have a trailing whitespace in one of your changed lines that needs to be removed to make the analyze task happy: https://cirrus-ci.com/task/5478008531451904

File: AUTHORS [hunk 2]
```diff
@@ -39,4 +39,5 @@ Marco Scannadinari <m@scannadinari.co.uk>
 Frederik Schweiger <mail@flschweiger.net>
 Martin Staadecker <machstg@gmail.com>
 Igor Katsuba <katsuba.igor@gmail.com>
-Diego Velásquez <diego.velasquez.lopez@gmail.com>
\ No newline at end of file
+Diego Velásquez <diego.velasquez.lopez@gmail.com>
+Simon Lightfoot <simon@devangels.london>
```
[the remark this answers is not in the document]
Comment (slightfoot) [reply AUTHORS]: There wasn't a newline before, so I decided not to add one. But no problem, I'll add this with the other changes.

**Comment by slightfoot** [comment]
@goderbauer review changes made.

**Review by goderbauer — APPROVED** [review]
LGTM

Thanks for your contribution! We can land this as soon as the tree is green again.

**Comment by goderbauer** [comment]
@slightfoot Looks like there's a merge conflict now. Can you resolve that please and then we can merge this!

**Comment by slightfoot** [comment]
@goderbauer merge conflict resolved. It was just the Authors file.

**Pull request merged**